src/lib/salad/hll.c

```diff
@@ -0,0 +1,1307 @@
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "hll.h"
+#include "trivia/util.h"
+#include "say.h"
+
+#include <stdlib.h>
+#include <string.h>
+#include <stddef.h>
+#include <limits.h>
+#include <math.h>
+#include <assert.h>
+
+/*
+ * The implementation of the algorithm is mainly based on the article
+ * HyperLogLog in Practice: Algorithmic Engineering of a State of The Art
+ * Cardinality Estimation Algorithm (https://research.google/pubs/pub40671/).
+ */
+
+enum HLL_CONSTANTS {
+	/*
+	 * 6 bits are needed to store the number of
+	 * trailing zeros of 64 bit hash.
+	 */
+	HLL_RANK_BITS = 6,
+
+	/* The maximum value that can be stored in HLL_RANK_BITS bits. */
+	HLL_RANK_MAX = (1 << HLL_RANK_BITS) - 1,
+
+	/* Number of bits stored in a register bucket. */
+	HLL_BUCKET_BITS = 24,
+	/* Number of bytes stored in a register bucket. */
+	HLL_BUCKET_BYTES = HLL_BUCKET_BITS / CHAR_BIT,
+
+	/*
+	 * Grow coefficient for the sparse representation. 2 allows to grow
+	 * to exact limit of the sparse representation
+	 * (equals to the memory usage of the dense representation).
+	 */
+	HLL_SPARSE_GROW_COEF = 2,
+
+	/*
+	 * The smallest precision of HyperLogLog that can start with the sparse
+	 * representation. Sparse representation for smaller precision can't
+	 * store more than 100 pairs so there is no need in this representation.
+	 */
+	HLL_SPARSE_MIN_PRECISION = 10,
+
+	/*
+	 * Initial size of sparsely represented HyperLogLog.
+	 * The initial size must accommodate at least a pairs header.
+	 */
+	HLL_SPARSE_INITIAL_BSIZE = 48,
+};
+
+/**
+ * Estimator that is used for the HyperLogLog algorithm.
+ * The algorithm allows to estimate number of distinct items using fixed amount
+ * of memory or even less. Memory requirements and estimation accuracy are
+ * determined by the algorithm precision parameter.
+ * For the dense representation the relative error is 1.04/sqrt(m) and the
+ * memory capacity is m*6 bits where m is number of counters wich
+ * equals to 2^precision.
+ * For the sparse representation the memory usage is proportional to the number
+ * of distinct elements that has added until it reaches the memory usage of the
+ * dense representation, and then it switches to the dense representation with
+ * fixed memory requretments. The sparse representation has the best accuracy.
+ */
+struct hll {
+	/** See the comment to HLL_REPRESENTATION enum. */
+	enum HLL_REPRESENTATION representation;
+	/**
+	 * Interpretation of the data depends on the representation.
+	 * For dense representation it's an array of registers of size
+	 * 2^precision * 6 bits. Registers store the maximum added rank of set
+	 * of hashes wich last precision bits are equal to register index.
+	 * For the sparse representation it's a sparsely represented
+	 * HyperLogLog. Instead of registers the sparse representation works
+	 * with pairs consisting of the rank and the index of the added hashes.
+	 */
+	uint8_t *data;
```

Comment:
Now it would be natural to use flexible array - `uin8_t data[]`. That would be simpler and more effective.